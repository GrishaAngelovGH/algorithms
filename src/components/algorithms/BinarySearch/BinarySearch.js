import { useState } from 'react'

import Alert from 'react-bootstrap/Alert'
import Button from 'react-bootstrap/Button'
import InputGroup from 'react-bootstrap/InputGroup'
import FormControl from 'react-bootstrap/FormControl'

import WaveBackground from '../../WaveBackground'

const BinarySearch = () => {
    const [elements, setElements] = useState([])
    const [criteria, setCriteria] = useState('')
    const [label, setLabel] = useState({ title: '', success: false })


    const handleCriteriaInputChange = ({ target: { value } }) => {
        setCriteria(parseInt(value))
    }

    const handleStepButtonClick = () => {
        const values = [...elements]

        if (!values.length || (values.length === 1 && values[0] !== parseInt(criteria))) {
            setLabel({ title: 'Element is not found!', success: false })
        }

        const index = Math.floor(values.length / 2)

        if (values[index] === parseInt(criteria)) {
            setLabel({ title: 'Element is found!', success: true })
        }

        const newElements = parseInt(criteria) < values[index] ? values.slice(0, index) : values.slice(index)

        setElements(newElements.sort((a, b) => a - b))
    }

    const handleRandomFillButtonClick = () => {
        const newElements = new Array(10).fill(0).map(() => Math.floor(Math.random() * 100)).filter(v => v > 0)

        setElements(newElements.sort((a, b) => a - b))
        setLabel({ title: '', success: false })
    }

    const handleResetButtonClick = () => {
        setElements([])
        setCriteria('')
        setLabel({ title: '', success: false })
    }

    return (
        <WaveBackground>
            <div className='row g-0 justify-content-center text-center p-3'>
                <div className='col-md-12'>
                    <h3 className='text-white'>Binary Search Algorithm</h3>

                    <div className='row justify-content-center'>
                        <div className='col-md-6'>
                            <InputGroup className='mb-3'>
                                <FormControl
                                    value={elements}
                                    disabled={true}
                                />
                                <Button onClick={handleRandomFillButtonClick}>
                                    Random Fill
                                </Button>
                            </InputGroup>

                            <InputGroup className='mb-3'>
                                <FormControl
                                    type='number'
                                    min={1}
                                    placeholder='Enter search criteria, ex: 3'
                                    value={criteria}
                                    onChange={handleCriteriaInputChange}
                                />
                            </InputGroup>

                            {
                                !label.title.length && (
                                    <Button onClick={handleStepButtonClick} disabled={criteria === ''}>
                                        Step
                                    </Button>
                                )
                            }

                            {
                                label.title.length > 0 && (
                                    <Button onClick={handleResetButtonClick}>
                                        Reset
                                    </Button>
                                )
                            }

                            {
                                elements.length > 0 && (
                                    <div className='row justify-content-center p-3'>
                                        {
                                            elements.map((v, i) => (
                                                <div key={i} className='col-md-1 mx-1 bg-success text-white rounded'>{v}</div>
                                            ))
                                        }
                                    </div>
                                )
                            }

                            {
                                label.title.length > 0 && (
                                    <Alert variant={label.success ? 'success' : 'danger'}>
                                        <Alert.Heading>{label.title}</Alert.Heading>
                                    </Alert>
                                )
                            }

                        </div>
                    </div>
                </div>
            </div>
        </WaveBackground>
    )
}

export default BinarySearch