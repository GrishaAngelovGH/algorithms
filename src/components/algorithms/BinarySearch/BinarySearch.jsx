import { useState, useEffect } from 'react'

import Alert from 'react-bootstrap/Alert'
import Button from 'react-bootstrap/Button'
import InputGroup from 'react-bootstrap/InputGroup'
import FormControl from 'react-bootstrap/FormControl'

import WaveBackground from 'components/WaveBackground'

const BinarySearch = () => {
	const [elements, setElements] = useState([])
	const [criteria, setCriteria] = useState('')
	const [middleIndex, setMiddleIndex] = useState(0)
	const [label, setLabel] = useState({ title: '', success: false })

	const handleCriteriaInputChange = ({ target: { valueAsNumber } }) => {
		setCriteria(valueAsNumber ? valueAsNumber : '')
	}

	const handleStepButtonClick = () => {
		const values = [...elements]

		if (!values.length || (values.length === 1 && values[0] !== criteria)) {
			setLabel({ title: 'Element is not found!', success: false })
		}

		if (values[middleIndex] === criteria) {
			setLabel({ title: 'Element is found!', success: true })
		}

		const newElements = criteria < values[middleIndex] ?
			values.slice(0, middleIndex) : values.slice(middleIndex)

		setElements(newElements)
	}

	const handleRandomFillButtonClick = () => {
		setElements(
			new Array(10)
				.fill(0)
				.map(() => Math.floor(Math.random() * 100))
				.filter(v => v > 0)
				.sort((a, b) => a - b)
		)

		setLabel({ title: '', success: false })
	}

	const handleResetButtonClick = () => {
		setElements([])
		setCriteria('')
		setLabel({ title: '', success: false })
	}


	useEffect(() => {
		setMiddleIndex(
			Math.floor(elements.length / 2)
		)
	}, [elements])

	return (
		<WaveBackground>
			<div className='row g-0 justify-content-center text-center p-3'>
				<div className='col-md-12'>
					<h3 className='text-white'>Binary Search Algorithm</h3>

					<div className='row justify-content-center'>
						<div className='col-md-6'>
							<InputGroup className='mb-3'>
								<FormControl value={elements} disabled={true} />

								<Button onClick={handleRandomFillButtonClick}>
									Random Fill
								</Button>
							</InputGroup>

							<InputGroup className='mb-3'>
								<FormControl
									min={1}
									type={'number'}
									value={criteria}
									placeholder={'Enter search criteria'}
									onChange={handleCriteriaInputChange}
								/>
							</InputGroup>

							<div className='row justify-content-center'>
								{
									elements.map((v, i) => {
										const middleElementClass = i === middleIndex ? 'border border-2 border-white' : ''
										const elementClass = `col-md-1 mx-1 bg-success text-white rounded ${middleElementClass}`

										return (
											<div key={i} className={elementClass}>{v}</div>
										)
									})
								}
							</div>

							<div className='row justify-content-center p-3'>
								<div className='col-md-2'>
									{
										!label.title.length && (
											<Button onClick={handleStepButtonClick} disabled={criteria === ''}>
												Step
											</Button>
										)
									}

									{
										label.title.length > 0 && (
											<Button onClick={handleResetButtonClick}>
												Reset
											</Button>
										)
									}
								</div>
							</div>

							{
								label.title.length > 0 && (
									<Alert variant={label.success ? 'success' : 'danger'}>
										<Alert.Heading>{label.title}</Alert.Heading>
									</Alert>
								)
							}
						</div>
					</div>
				</div>
			</div>
		</WaveBackground>
	)
}

export default BinarySearch