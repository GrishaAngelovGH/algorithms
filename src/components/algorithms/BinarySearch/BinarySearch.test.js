import { render } from '@testing-library/react'

import BinarySearch from './BinarySearch'

test('should render component', () => {
    const container = render(<BinarySearch />)

    expect(container).toMatchSnapshot()
})