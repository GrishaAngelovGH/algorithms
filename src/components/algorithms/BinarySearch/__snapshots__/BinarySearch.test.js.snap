// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="waveWrapper waveAnimation"
      >
        <div
          class="waveWrapperInner bgTop"
        >
          <div
            class="wave waveTop"
          />
        </div>
        <div
          class="waveWrapperInner bgMiddle"
        >
          <div
            class="wave waveMiddle"
          />
        </div>
        <div
          class="waveWrapperInner bgBottom"
        >
          <div
            class="wave waveBottom"
          />
        </div>
      </div>
      <div
        class="wave-content"
      >
        <div
          class="row g-0 justify-content-center text-center p-3"
        >
          <div
            class="col-md-12"
          >
            <h3
              class="text-white"
            >
              Binary Search Algorithm
            </h3>
            <div
              class="row justify-content-center"
            >
              <div
                class="col-md-6"
              >
                <div
                  class="mb-3 input-group"
                >
                  <input
                    class="form-control"
                    disabled=""
                    value=""
                  />
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Random Fill
                  </button>
                </div>
                <div
                  class="mb-3 input-group"
                >
                  <input
                    class="form-control"
                    min="1"
                    placeholder="Enter search criteria"
                    type="number"
                    value=""
                  />
                </div>
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="button"
                >
                  Step
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="waveWrapper waveAnimation"
    >
      <div
        class="waveWrapperInner bgTop"
      >
        <div
          class="wave waveTop"
        />
      </div>
      <div
        class="waveWrapperInner bgMiddle"
      >
        <div
          class="wave waveMiddle"
        />
      </div>
      <div
        class="waveWrapperInner bgBottom"
      >
        <div
          class="wave waveBottom"
        />
      </div>
    </div>
    <div
      class="wave-content"
    >
      <div
        class="row g-0 justify-content-center text-center p-3"
      >
        <div
          class="col-md-12"
        >
          <h3
            class="text-white"
          >
            Binary Search Algorithm
          </h3>
          <div
            class="row justify-content-center"
          >
            <div
              class="col-md-6"
            >
              <div
                class="mb-3 input-group"
              >
                <input
                  class="form-control"
                  disabled=""
                  value=""
                />
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Random Fill
                </button>
              </div>
              <div
                class="mb-3 input-group"
              >
                <input
                  class="form-control"
                  min="1"
                  placeholder="Enter search criteria"
                  type="number"
                  value=""
                />
              </div>
              <button
                class="btn btn-primary"
                disabled=""
                type="button"
              >
                Step
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
